src/core/execute.c

```diff
@@ -3374,6 +3383,30 @@ static int exec_child(
                 }
         }
 
+        if (needs_sandboxing) {
+#if HAVE_SELINUX
+                if (use_selinux && params->selinux_context_net && socket_fd >= 0) {
+                        r = mac_selinux_get_child_mls_label(socket_fd, command->path, context->selinux_context, &mac_selinux_context_net);
+                        if (r < 0) {
+                                *exit_status = EXIT_SELINUX_CONTEXT;
+                                return log_unit_error_errno(unit, r, "Failed to determine SELinux context: %m");
+                        }
+                }
+#endif
+
+                /* If we're unprivileged, set up the user namespace first to enable use of the other namespaces.
+                 * Users with CAP_SYS_ADMIN can set up user namespaces last because they will be able to
+                 * set up the all of the other namespaces (i.e. network, mount, UTS) without a user namespace. */
+                if (context->private_users && !have_effective_cap(CAP_SYS_ADMIN)) {
+                        userns_setup = true;
+                        r = setup_private_users(uid, gid);
```

Comment:
hmm, this means we now switch userns before dropping groups, see enforce_groups() below. This doesn't look correct to me. We should enforce the groups in the host's view and then switch userns after, no?

Reply:
Re: The selinux part, I don't think it really matters whether it's before or after the user ns setup but I have a slight preference right now to keep it before as that's how it was ordered originally.

Re: Groups, I was testing with non-root that did not have `CAP_SETGID` so the current ordering would work fine (because it would fail either way to change groups or set up the user ns mapping). But you're right that in the case where non-root users had `CAP_SETGID` then `enforce_groups` would need to be done before switching user ns. So I'll move those around.

Reply:
Actually maybe I take that back... Even if a non-root user had `CAP_SETGID` and did enforce groups before setting user namespaces, they would not be able to write to `/proc/ppid/setgroups` (and mappings) because the process would not be dumpable after changing effective group IDs. But in the current layout it can still set mappings and change group IDs.

But seeing as how this is the unprivileged case, and `User=/Group=` is documented to not be supported by user services anyways, maybe that's what we want? Enforce groups in the host view and not allow user ns in that case?

Reply:
Please add a test (in TEST/ somewhere?) for this. This is complex enough that merging it without any tests shouldn't be done.